src/Compilers/CSharp/Test/Syntax/Parsing/DeclarationExpressionTests.cs

```diff
@@ -0,0 +1,1186 @@
+                }
+            }
+            EOF();
+        }
+
+        [Fact]
+        public void NewForeach_06()
+        {
+            UsingStatement("foreach ((int x, var (y, z)) in e) {}");
+            N(SyntaxKind.ForEachVariableStatement);
+            {
+                N(SyntaxKind.ForEachKeyword);
+                N(SyntaxKind.OpenParenToken);
+                N(SyntaxKind.TupleExpression);
+                {
+                    N(SyntaxKind.OpenParenToken);
+                    N(SyntaxKind.Argument);
+                    {
+                        N(SyntaxKind.DeclarationExpression);
+                        {
+                            N(SyntaxKind.PredefinedType);
+                            {
+                                N(SyntaxKind.IntKeyword);
+                            }
+                            N(SyntaxKind.SingleVariableDesignation);
+                            {
+                                N(SyntaxKind.IdentifierToken, "x");
+                            }
+                        }
+                    }
+                    N(SyntaxKind.CommaToken);
+                    N(SyntaxKind.Argument);
+                    {
+                        N(SyntaxKind.DeclarationExpression);
+                        {
+                            N(SyntaxKind.IdentifierName);
+                            {
+                                N(SyntaxKind.IdentifierToken, "var");
+                            }
+                            N(SyntaxKind.ParenthesizedVariableDesignation);
+                            {
+                                N(SyntaxKind.OpenParenToken);
+                                N(SyntaxKind.SingleVariableDesignation);
+                                {
+                                    N(SyntaxKind.IdentifierToken, "y");
+                                }
+                                N(SyntaxKind.CommaToken);
+                                N(SyntaxKind.SingleVariableDesignation);
+                                {
+                                    N(SyntaxKind.IdentifierToken, "z");
+                                }
+                                N(SyntaxKind.CloseParenToken);
+                            }
+                        }
+                    }
+                    N(SyntaxKind.CloseParenToken);
+                }
+                N(SyntaxKind.InKeyword);
+                N(SyntaxKind.IdentifierName);
+                {
+                    N(SyntaxKind.IdentifierToken, "e");
+                }
+                N(SyntaxKind.CloseParenToken);
+                N(SyntaxKind.Block);
+                {
+                    N(SyntaxKind.OpenBraceToken);
+                    N(SyntaxKind.CloseBraceToken);
+                }
+            }
+            EOF();
+        }
+
+        [Fact]
+        public void NewForeach_07()
+        {
+            // PROTOTYPE(wildcards): SHOULD BE some kind of *semantic* error here
```

Comment:
If you want this PR to target `master` branch (for RC3), don't forget to replace the PROTOTYPE markers with issues.